addons/netfox.internals/bimap.gd

```diff
@@ -0,0 +1,52 @@
+extends RefCounted
+class_name _BiMap
+
+# Maps one-to-one associations in a bidirectional way
+
+# TODO: Tests
+
+var _keys_to_values := {}
+var _values_to_keys := {}
+
+func put(key: Variant, value: Variant) -> void:
+	_keys_to_values[key] = value
+	_values_to_keys[value] = key
```

Comment:
This breaks as soon as two values have the same key. There has to be a check to guarantee the bidirectional mapping.

Reply:
That is not the intended use. This class is for one-to-one relations as mentioned at the top of the file.

Reply:
I still would add an `assert` to make sure this invariant is upheld at least in debug builds.

Reply:
I still consider this as standard behavior for the data structure. Same way the dictionary doesn't warn you when you're trying to associate multiple values with the same key - it just overwrites the association. The same happens here.

Granted, I'm willing to add an assert or warning down the line in case we run into a situation where this behavior gets confusing, but I do not see the point at this time.

Reply:
If it would only overwrite the old data I would be fine with it. But in this unfortunate cases we end up in an inconsistent state.
For example, we first add a mapping from a -> b, then a mapping from c -> b.
_keys_to_values is then { a -> b, c -> b} and _values_to_keys is { b -> c }

Reply:
Would have helped if you started with that example :smile: That was fixed a bit before you posted this comment, but let me know if the tests don't cover that case.